### PR TITLE
Adding DecodeScaled method.

### PR DESCRIPTION
**Pull request by jamesshoebox**
Uses the advanced decoding API to scale an image while decoding. Saves a lot of
memory and CPU time compared to decoding and then resizing.

Our use case: load original size WEBP -> resize -> encode to JPEG

Old method: `webp.Decode` -> `nfnt/resize` -> `jpeg.Encode`  
New method: `webp.DecodeScaled` -> `jpeg.Encode`

Benchmarks, with source file 3264x2448px, output file 256x192px:
```
BenchmarkMediumOld-8   	       5	 236803936 ns/op
BenchmarkMediumNew-8   	      30	  48662007 ns/op
```

### CODE REVIEW

**Comment by chai2010** [comment]
I think the `DecodeRGBA***` is the better name than `DecodeScaled`,
same as `DecodeRGBA` and `DecodeGray`.

```
func DecodeRGBAToSize(data []byte, width, height int) (m *image.RGBA, err error);
```

then we can add simalir api for Gray and RGB:

```
func DecodeGrayToSize(data []byte, width, height int) (m *image.Gray, err error);
func DecodeRGBToSize(data []byte, width, height int) (m *RGBImage, err error);
```

Thanks for your work.

**Comment by chai2010** [comment]
I think the `DecodeRGBA***` is the better name than `DecodeScaled`,
same as `DecodeRGBA` and `DecodeGray`.

```
func DecodeRGBAToSize(data []byte, width, height int) (m *image.RGBA, err error);
```

then we can add simalir api for Gray and RGB:

```
func DecodeGrayToSize(data []byte, width, height int) (m *image.Gray, err error);
func DecodeRGBToSize(data []byte, width, height int) (m *RGBImage, err error);
```

Thanks for your work.

**Comment by jamesshoebox** [comment]
That sounds good to me. I made those updates, let me know what you think. Thanks!

File: capi.go [hunk 1]
```diff
@@ -109,34 +109,30 @@ func webpDecodeRGBA(data []byte) (pix []byte, width, height int, err error) {
...
 
-func webpDecodeScaled(data []byte, width, height int) (pix []byte, err error) {
-	if len(data) == 0 {
-		err = errors.New("webpDecodeScaled: bad arguments")
-		return
+func webpDecodeGrayToSize(data []byte, width, height int) (pix []byte, err error) {
+	pix = make([]byte, int(width*height))
+	res := C.webpDecodeGrayToSize((*C.uint8_t)(unsafe.Pointer(&data[0])), C.size_t(len(data)), C.int(width), C.int(height), (*C.uint8_t)(unsafe.Pointer(&pix[0])), C.size_t(len(pix)))
+	if res != C.VP8_STATUS_OK {
+		err = errors.New("webpDecodeGrayToSize: failed")
```
Comment (chai2010) [inline capi.go]: `if err != nil`, need set pix to `nil`.

File: capi.go [hunk 2]
```diff
@@ -109,34 +109,30 @@ func webpDecodeRGBA(data []byte) (pix []byte, width, height int, err error) {
...
-	rgba.stride = C.int(4 * width)
-	rgba.size = C.size_t(len(pix))
-
-	res := C.WebPDecode((*C.uint8_t)(unsafe.Pointer(&data[0])), C.size_t(len(data)), &config)
+func webpDecodeRGBToSize(data []byte, width, height int) (pix []byte, err error) {
+	pix = make([]byte, int(3*width*height))
+	res := C.webpDecodeRGBToSize((*C.uint8_t)(unsafe.Pointer(&data[0])), C.size_t(len(data)), C.int(width), C.int(height), (*C.uint8_t)(unsafe.Pointer(&pix[0])), C.size_t(len(pix)))
 	if res != C.VP8_STATUS_OK {
-		err = errors.New("webpDecodeScaled: failed")
+		err = errors.New("webpDecodeRGBToSize: failed")
```
Comment (chai2010) [inline capi.go]: ditto

File: capi.go [hunk 3]
```diff
@@ -109,34 +109,30 @@ func webpDecodeRGBA(data []byte) (pix []byte, width, height int, err error) {
...
 	}
+	return
+}
 
+func webpDecodeRGBAToSize(data []byte, width, height int) (pix []byte, err error) {
+	pix = make([]byte, int(4*width*height))
+	res := C.webpDecodeRGBAToSize((*C.uint8_t)(unsafe.Pointer(&data[0])), C.size_t(len(data)), C.int(width), C.int(height), (*C.uint8_t)(unsafe.Pointer(&pix[0])), C.size_t(len(pix)))
+	if res != C.VP8_STATUS_OK {
+		err = errors.New("webpDecodeRGBAToSize: failed")
```
Comment (chai2010) [inline capi.go]: ditto

File: internal/libwebp/include/webp.h [hunk 4]
```diff
@@ -32,6 +32,19 @@ uint8_t* webpDecodeRGBA(
 	int* width, int* height
 );
 
+int webpDecodeGrayToSize(const uint8_t* data, size_t data_size,
+	int width, int height,
+	uint8_t* gray, size_t gray_size
```
Comment (chai2010) [inline internal/libwebp/include/webp.h]: I think the `gray_size` is not needed.
I suggest a `widthStep` param, then we can use `webpDecodeGrayToSize` to decode a subimage to a large image buffer.

File: internal/libwebp/include/webp.h [hunk 5]
```diff
@@ -32,6 +32,19 @@ uint8_t* webpDecodeRGBA(
...
+);
+int webpDecodeRGBToSize(const uint8_t* data, size_t data_size,
+	int width, int height,
+	uint8_t* rgb, size_t rgb_size
```
Comment (chai2010) [inline internal/libwebp/include/webp.h]: ditto

File: internal/libwebp/include/webp.h [hunk 6]
```diff
@@ -32,6 +32,19 @@ uint8_t* webpDecodeRGBA(
...
+);
+int webpDecodeRGBAToSize(const uint8_t* data, size_t data_size,
+	int width, int height,
+	uint8_t* rgba, size_t rgba_size
```
Comment (chai2010) [inline internal/libwebp/include/webp.h]: ditto

File: internal/libwebp/src/webp.c [hunk 7]
```diff
@@ -87,6 +87,97 @@ uint8_t* webpDecodeRGBA(
 	return WebPDecodeRGBA(data, data_size, width, height);
 }
 
+int webpDecodeGrayToSize(const uint8_t* data, size_t data_size,
+	int width, int height,
+	uint8_t* gray, size_t gray_size
+) {
+	if (gray_size < width * height) {
+		return -1;
```
Comment (chai2010) [inline internal/libwebp/src/webp.c]: i think the panic is the better choise, so `gray_size` is not needed.

File: internal/libwebp/src/webp.c [hunk 8]
```diff
@@ -87,6 +87,97 @@ uint8_t* webpDecodeRGBA(
...
+	}
+
+	size_t stride = config.output.u.YUVA.y_stride;
+	uint8_t* src = config.output.u.YUVA.y;
+	uint8_t* dst = gray;
+	int i;
+
+	if(stride == width && config.output.u.YUVA.y_size == gray_size) {
+		memcpy(dst, src, gray_size);
+	}
```
Comment (chai2010) [inline internal/libwebp/src/webp.c]: keep the code format same as `go fmt`

File: internal/libwebp/src/webp.c [hunk 9]
```diff
@@ -87,6 +87,97 @@ uint8_t* webpDecodeRGBA(
...
+	else {
+		for(i = 0; i < height; ++i) {
+			memmove(dst, src, width);
+			src += stride;
+			dst += width;
```
Comment (chai2010) [inline internal/libwebp/src/webp.c]: `dst += width`, use `dstStride` or name `widthStep` is better:
`dst += widthStep`

**Comment by chai2010** [comment]
Seems good to me.

some other suggest:

1. add test and bench for new API
2. can you help me to improve the `webpEncodeGray` (it use `WebPEncodeRGB`), the best way is use YUV style encode API(may be in new PR).

Thanks.

**Comment by jamesshoebox** [comment]
Thanks for the feedback. I made these changes. Also, in a separate commit, I added a large lossy image to better show the benchmark improvements. I don't know if you want to include such a large file in the repo.

The results on my machine are:
```
$ go test -bench photo_lossy
testing: warning: no tests to run
PASS
BenchmarkDecode_photo_lossy_chai2010_webp-8       	      20	  89326742 ns/op
BenchmarkDecode_photo_lossy_x_image_webp-8        	       5	 265046500 ns/op
BenchmarkDecode_photo_lossy_chai2010_webp_tosize-8	      20	  63967415 ns/op
ok  	github.com/chai2010/webp/bench	6.565s
```

We deployed this change to our production environment and are seeing even more dramatic improvements there, due to the elimination of a separate resize operation. For example, average times to resize this `photo.lossy.webp` image to 256x192px dropped from about 900 ms to about 200 ms.

**Comment by jamesshoebox** [comment]
BTW, don't mind helping with `webpEncodeGray` but would prefer to do that separately

**Pull request merged**